src/Services/Processor/Substrate/BlockProcessor.php

```diff
@@ -283,10 +283,24 @@ protected function setBlockExtrinsic(Substrate $blockchain, Block $block): Block
         return $block;
     }
 
+    protected function waitIfEmpty(int $try, int $blockNumber, string $action): void
+    {
+        if ($try > 0) {
+            usleep($sleep = 250000 * $try ** 2);
+            $this->warn(sprintf('Retrying to fetch %s for block #%s in %s seconds', $action, $blockNumber, $sleep / 1000000));
+        }
+    }
+
     protected function fetchEvents(Block $block): Block
     {
         $syncTime = now();
-        $block = $this->setBlockEvent($this->persistedClient, $block);
+        $try = 0;
+
+        while (empty($block->events)) {
```

Comment:
This may cause indefinite loop if the condition is not met

Reply:
I'm aware and the condition NEED to be met. All blocks have events, all blocks have extrinsics, if the condition is not met, means the data is wrong, means it should not proceed anyway.